### PR TITLE
Add FileInput component

### PR DESCRIPTION
**Pull request by samshara**

### CODE REVIEW

File: src/components/FileInput/index.tsx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            }
+        }, [onChange, isValidFile, accept, getStatus],
+    );
+
+    const handleDrop: React.DragEventHandler<HTMLDivElement> = useCallback((e) => {
+        e.preventDefault();
+        const fileList = Array.from(e.dataTransfer.files);
+        const validFiles = fileList.filter((f) => isValidFile(f.name, f.type, accept));
+        if (!multiple && fileList.length > 1) {
+            setStatus('Multiple file selection no allowed. Please, select a single file.');
```
Comment (tnagorra) [inline src/components/FileInput/index.tsx]:
```suggestion
            setStatus('Multiple file selection not allowed. Please select a single file.');
```

File: src/components/FileInput/index.tsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        dragEnterRef.current += 1;
+    }, [setIsBeingDraggedOver]);
+
+    const handleDragLeave = useCallback(() => {
+        dragEnterRef.current -= 1;
+
+        if (dragEnterRef.current === 0) {
+            setIsBeingDraggedOver(false);
+        }
+    }, [setIsBeingDraggedOver]);
```
Comment (tnagorra) [inline src/components/FileInput/index.tsx]: The drag handling should be reused from DropContainer.
Maybe we need to create a useDrop hook

File: src/components/FileInput/index.tsx (hunk 1 above)
Comment (tnagorra) [reply src/components/FileInput/index.tsx]: Why not just take the first file and ignore the rest?

File: src/components/FileInput/index.tsx [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+    const getStatus = useCallback((files: File[]) => {
+        if (files && files.length > 1) {
+            return `${files?.length} files selected`;
+        }
+        if (files && files.length === 0) {
+            return 'Invalid files selected';
+        }
```
Comment (tnagorra) [inline src/components/FileInput/index.tsx]: Wouldn't this also be no file chosen?

**Review by tnagorra — CHANGES_REQUESTED** [review]
FileInput is not exported!

File: src/components/FileInput/index.tsx (hunk 1 above)
Comment (samshara) [reply src/components/FileInput/index.tsx]: we can select arbitrary file types when drag and dropping files, so we must check each one of them.

File: src/components/FileInput/index.tsx (hunk 2 above)
Comment (samshara) [reply src/components/FileInput/index.tsx]: The DropContainer only works for text.

File: src/components/FileInput/index.tsx (hunk 1 above)
Comment (tnagorra) [reply src/components/FileInput/index.tsx]: I mean, we can just take the first valid file when user selects multiple files during drag and drop.

File: src/components/FileInput/index.tsx (hunk 2 above)
Comment (tnagorra) [reply src/components/FileInput/index.tsx]: But we can create a useDrop hook that's more generic that can be used on both cases.

File: src/components/FileInput/index.tsx [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        if (files && files.length > 1) {
+            return `${files?.length} files selected`;
+        }
+        if (files && files.length === 0) {
+            return 'No file chosen';
+        }
+        if (files) {
+            return files[0]?.name;
+        }
+        return 'No file chosen';
```
Comment (tnagorra) [inline src/components/FileInput/index.tsx]:
```suggestion
        if (!files || files.length === 0) {
            return 'No file chosen';
        }
        if (files.length > 1) {
            return `${files?.length} files selected`;
        }
        return files[0]?.name;
```

File: src/components/FileInput/index.tsx [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        }, [onChange, isValidFile, accept, getStatus],
+    );
+
+    const handleDrop: React.DragEventHandler<HTMLDivElement> = useCallback((e) => {
+        e.preventDefault();
+        const fileList = Array.from(e.dataTransfer.files);
+        const validFiles = fileList.filter((f) => isValidFile(f.name, f.type, accept));
+        if (!multiple && fileList.length > 1) {
+            setStatus('Multiple file selection not allowed. Please, select a single file.');
+        } else {
```
Comment (tnagorra) [inline src/components/FileInput/index.tsx]: We should add this check to handleChange as well (to be 100% certain)

File: src/components/FileInput/index.tsx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        isValidFile,
+    ]);
+
+    const {
+        dropping,
+        onDragOver,
+        onDragEnter,
+        onDragLeave,
+        onDrop,
+    } = useDropHandler(handleDrop);
```
Comment (tnagorra) [inline src/components/FileInput/index.tsx:121]: Did we already have a useDropHandler?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/components/FileInput/index.tsx (hunk 1 above)
Comment (samshara) [reply src/components/FileInput/index.tsx]: so only a single file should be selected during drag and drop?

File: src/components/FileInput/index.tsx (hunk 6 above)
Comment (samshara) [reply src/components/FileInput/index.tsx:121]: yes we did.

File: src/components/FileInput/index.tsx (hunk 5 above)
Comment (samshara) [reply src/components/FileInput/index.tsx]: does't the ```multiple``` prop passed to ```<input type='file'>``` control number of files that can be selected?. If so I think the condition won't arise in handleChange.

File: src/components/FileInput/index.tsx (hunk 1 above)
Comment (tnagorra) [reply src/components/FileInput/index.tsx]: Yes. I think that's the correct behavior when user tries to drag drop a lot of elements when only one is allowed

File: src/components/FileInput/index.tsx (hunk 5 above)
Comment (tnagorra) [reply src/components/FileInput/index.tsx]: It should. But, it just depends on each browser. Better be safe than sorry!

File: src/components/FileInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+        const fileTypeList = acceptString.split(/,\s+/);
+        return fileTypeList.some((fileType) => {
+            // check mimeType such as image/png or image/*
+            if (mimeType === fileType || (!!mimeMatch && `${mimeMatch[0]}*` === fileType)) {
+                return true;
+            }
+            return !!extensionMatch && extensionMatch[0].toLowerCase() === fileType.toLowerCase();
+        });
+    }, []);
```
Comment (tnagorra) [inline src/components/FileInput/index.tsx]: We should move this outside the react component. (maybe export it as well)
One other added advantage is that we don't need to add this to dependency for other hooks

**Review by tnagorra — APPROVED** [review]

**Pull request merged**